nuxeo-runtime/nuxeo-connect-standalone/src/main/java/org/nuxeo/launcher/connect/ConnectBroker.java

```diff
@@ -855,14 +871,57 @@ public boolean pkgInstall(List<String> packageIdsToInstall) {
      */
     public boolean pkgInstall(List<String> packageIdsToInstall, boolean ignoreMissing) {
         log.debug("Installing: " + packageIdsToInstall);
-        for (String pkgId : packageIdsToInstall) {
+        Queue<String> remaining = new LinkedList<>(packageIdsToInstall);
+        while (!remaining.isEmpty()) {
+            String pkgId = remaining.poll();
             if (pkgInstall(pkgId, ignoreMissing) == null && !ignoreMissing) {
                 return false;
             }
+            if (isLauncherChanged()) {
+                // TODO Handle ignoreMissing?
+                persistPendingCommand(CommandInfo.CMD_INSTALL, remaining);
+                throw new LauncherRestartException();
+            }
         }
         return true;
     }
 
+    /**
+     * Persists the pending package operation into file system. It's useful when Nuxeo launcher is about to exit.
+     * <p>
+     * The given commands will be appended as lines to the target file, {@link #INSTALL_AFTER_RESTART}.
+     *
+     * @param commands commands to persist, each string is an individual command
+     * @throws IllegalStateException if any exception occurs
+     * @see #INSTALL_AFTER_RESTART
+     * @since 10.2
+     */
+    protected void persistPendingCommands(Iterable<String> commands) {
+        Path path = getInstallAfterRestartPath();
+        try {
+            Files.write(path, commands, StandardOpenOption.CREATE, StandardOpenOption.APPEND);
+        } catch (IOException e) {
+            throw new IllegalStateException("Cannot write to file " + path, e);
+        }
+    }
+
+    /**
+     * @see #persistPendingCommands(Iterable)
+     * @since 10.2
+     */
+    protected void persistPendingCommand(String commandName, Collection<String> args) {
+        if (args.isEmpty()) {
+            return;
+        }
+        Path path = getInstallAfterRestartPath();
+        String line = commandName + ' ' + String.join(" ", args);
```

Comment:
to verify (spaces in path): `install /path/to/some package.zip /another path/to/a/package.zip`